### PR TITLE
udev: make algorithm that selects highest priority devlink less susceptible to race conditions

### PR DESCRIPTION
**Pull request by msekletar**

### CODE REVIEW

File: src/udev/udev-node.c [hunk 1]
```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
-                        (void) rmdir_parents(slink, "/");
-        } else
-                (void) node_symlink(dev, target, slink);
-
         if (add)
                 do {
                         _cleanup_close_ int fd = -1;
 
-                        r = mkdir_parents(filename, 0755);
+                        r = mkdir_parents_label(filename, 0755);
```
Comment (poettering) [inline src/udev/udev-node.c]: hmm, any particular reason for this? maybe split this into separate commit

**Comment by msekletar** [comment]
@mwilck I rebased your `udev-test.pl` work on top of current master implemented different algorithm to avoid race-condition. Also thanks to @poettering for all the suggestions!

File: src/udev/udev-event.c [hunk 2]
```diff
@@ -1023,10 +1023,6 @@ int udev_event_execute_rules(UdevEvent *event,
         if (r < 0)
                 return r;
 
-        r = update_devnode(event);
-        if (r < 0)
-                return r;
-
```
Comment (poettering) [inline src/udev/udev-event.c]: hmm, I'd really keep this here in place, so that we give the guarantee that the device node is fully set up once we update the db. That's a really important property: the udev db should only be updated once all the other stuff is in place (in particular all the metadata the db carries, and device node info is among that), so that if people enumerate things with the udev db the data stored in there actually matches reality.

Could you explain why you moved this down? given that the symlinks db is separate from the regular device db there should not be a need to move this down?

File: src/udev/udev-node.c [hunk 3]
```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+
+                if (add) {
+                        if (stat(dirname, &st2) < 0)
+                                return -errno;
+
+                        if (!stat_inode_unmodified(&st1, &st2))
+                                continue;
+                }
```
Comment (poettering) [inline src/udev/udev-node.c]: this if block should be moved down, to the end of the loop iteration, i.e. after the unlink/node_symlink() call. i.e. if two workers run this code in parallel,  one one does the link_find_prioritized first, and gets rescheduled before it can issue the node_symlink() and another worker then gets shceduled and manages to go through both, then the first worker will update the symlink with out of date info once it is scheduled again. 

Hence: the mtime "bracket" should cover both parts: the determination of the priorized symlink and the creation of the symlink. i.e. the first stat() should be at the beginning of the loop iteration, the second at the very end of the loop iteration, and if mtimes still match the loop should be exited.

(oh, and we can optimize the stat thing here: move the first stat() before the loop, and then use the second stat()'s data as the first stat()'s data of the next loop iteration, if you follow what I mean? stat()s are cheap but not free, and in the contended case this would half the number of stat()s we need)

File: src/udev/udev-node.c [hunk 4]
```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+                                (void) rmdir_parents(slink, "/");
+
+                        break;
+                } else {
+                        r = node_symlink(dev, target, slink);
+                        if (r < 0) {
+                                (void) unlink(filename);
+                                break;
+                        } else if (r == 1)
+                                continue;
```
Comment (poettering) [inline src/udev/udev-node.c]: why this check? if you remove it, then the code will be exactly the same, as this is at the end of the loop iteration anyway

File: src/udev/udev-node.c [hunk 5]
```diff
@@ -212,26 +216,50 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
-                        r = mkdir_parents(filename, 0755);
+                        r = mkdir_parents_label(filename, 0755);
                         if (!IN_SET(r, 0, -ENOENT))
                                 break;
                         fd = open(filename, O_WRONLY|O_CREAT|O_CLOEXEC|O_TRUNC|O_NOFOLLOW, 0444);
                         if (fd < 0)
                                 r = -errno;
                 } while (r == -ENOENT);
 
+        for (int i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
```
Comment (poettering) [inline src/udev/udev-node.c]: there should probably be some error returned if the retry limit was reached

File: src/udev/udev-node.c (hunk 4 above)
Comment (poettering) [reply src/udev/udev-node.c]: if you move second stat() check to the end of the loop iteration then the primary exit condition for the loop should be that mtime still matches. that's the key really.

**Review by poettering — CHANGES_REQUESTED** [review]
I love where this is going (obviously), some comments though

File: src/udev/udev-node.c (hunk 1 above)
Comment (msekletar) [reply src/udev/udev-node.c]: Probably there is no need for that. Parent db directory (`/run/udev/links`) has the correct label and child (links) directories should inherit the label automatically.

File: src/udev/udev-event.c (hunk 2 above)
Comment (msekletar) [reply src/udev/udev-event.c]: The idea is that `link_find_prioritized()` gets information about link priorities of other devices that claim the same symlink from their database entires. By changing the order we increase the chance that it finds a device with highest priority. Btw, if I don't do this `udev-test.pl` starts failing on test cases that simulate parallel addition of the devices that all want the same symlink (with different priority).

As for guarantees, well, once you receive uevent the db is updated already. Even previously, you could get into state where db entry contained data that didn't match the reality since symlinks where updated before db. I think the only guarantee is that for single kernel event, once processed and client received device event, the DB is consistent. This change doesn't alter that guarantee.

File: src/udev/udev-node.c (hunk 3 above)
Comment (msekletar) [reply src/udev/udev-node.c]: > this if block should be moved down, to the end of the loop iteration, i.e. after the unlink/node_symlink() call. i.e. if two workers run this code in parallel, one one does the link_find_prioritized first, and gets rescheduled before it can issue the node_symlink() and another worker then gets shceduled and manages to go through both, then the first worker will update the symlink with out of date info once it is scheduled again.

This is handled by checking return value of `node_symlink()`. We can't rely only on mtime of symlink directory.

File: src/udev/udev-node.c (hunk 4 above)
Comment (msekletar) [reply src/udev/udev-node.c]: Good point, there used to be some code bellow and I didn't fix the else branch

File: src/udev/udev-node.c (hunk 1 above)
Comment (poettering) [reply src/udev/udev-node.c]: i'd leave this out unless there's a real reason to add it, i.e. a real bug or complaint about an incorrect label

File: src/udev/udev-event.c (hunk 2 above)
Comment (poettering) [reply src/udev/udev-event.c]: still, the udev db should reflect not report symlinks that don't exist yet. the udev db should be updated last.

The symlink db is kep separate and is maintained inside of update_devnode(), hence the udev db should not be relevant to create the correct symlink.

not sure i grok the udev-test.pl failure, that's something to track down.

udev is not just about delivering uevents. It's also that when you enumerate stuff (i.e. udev_enumerate, not udev_monitor) things you discover then are already in place.

Thus, the order of the three things should be: first update symlink/devnode, second update the db, third tell everyone about it

File: src/udev/udev-node.c (hunk 3 above)
Comment (poettering) [reply src/udev/udev-node.c]: the return code of node_symlink() has no effect in the current patch though: it just decides whether to use "continue" as last thing in the loop iteration or not, but that should not make any difference, see below.

why can't we rely on the mtime of the symlink dir?

Comment (msekletar) [reply src/udev/udev-node.c]: Because the algorithm as designed can still pick the wrong symlink, the race is still fundamentally there. If it does and it overwrites higher priority symlink (we didn't see that it is supposed to be higher priority when calling link_find_pioritized()) we should enter next loop iteration.

Comment (msekletar) [reply src/udev/udev-node.c]: I think the key here is that update of symlink dir and actual /dev links is done at different times.

Comment (msekletar) [reply src/udev/udev-node.c]: And update of db entry that we use to actually determine highest priority symlink happens at yet another time.

Comment (poettering) [reply src/udev/udev-node.c]: it shouldn't matter if we pick up the wrong symlink momentarily, as long as we recover immediately. And we do: because we'll redo things immediately if the symlink db changed while we created the symlink.

As I said before: the symlink will not be totally off, it will just be "slightly" off, i.e. pointing to the next highest symlink or so, i.e. still to a device with the correct identifying data. And it will be corrected immediately.

**Comment by msekletar** [comment]
Updated. @poettering PTAL.

As agreed offline we now run algorithm (`update_devnode()`) twice, first time we may not be successful but we will correct symlinks momentarily because we run second time after the udev db is fully updated.

**Comment by mwilck** [comment]
Hi Michael,

I finally came down to testing your code. First I made sure the issue reproduced with current master (it does, easily). Then I tested with 50 devices (`python sd-test.py . 50`), and it worked well :+1: . With 200 devices, it also succeeds, but we got scaling issues again. 

`perf top` shows output like this:

```
+   70,32%     0,00%  test-udev                 [.] link_update                                                                       
-   69,95%     0,11%  test-udev                 [.] link_find_prioritized                                                             
+   47,92%     1,35%  [kernel]                  [k] entry_SYSCALL_64_after_hwframe                                                    
+   47,91%     0,07%  libsystemd-shared-246.so  [.] sd_device_new_from_device_id                                                      
+   47,07%     0,13%  libsystemd-shared-246.so  [.] sd_device_new_from_devnum                                                         
+   46,89%     0,04%  libsystemd-shared-246.so  [.] sd_device_new_from_syspath                                                        
+   46,67%     1,20%  [kernel]                  [k] do_syscall_64                                                                     
+   44,15%     0,12%  libsystemd-shared-246.so  [.] device_set_syspath                                                                
```

IIRC this is not worse than what I saw previously with my code - hard to compare anyway because the hardware has changed in the meantime.

Anyway, great job!

**Comment by mwilck** [comment]
Some numbers from my test system (24 x Intel(R) Xeon(R) CPU E5-2620 v3 @ 2.40GHz, 64 GB RAM), from `time udev-test.pl 161`, all times in seconds:

number of devices | real | user | sys
-------------------------|----|----|---
1 | 1.3 | 5.1 | 4.5
10 | 3.0 | 24 | 27 
20 | 6.5 | 57 | 64
50 | 24 | 220 | 250
100 | 71 | 690 | 760
200 | 300 | 3000 | 3300

**Comment by msekletar** [comment]
@mwilck thank you very much for testing!

Btw, the alternative locking scheme (POSIX semaphores or we could use pthread_mutex) in memfd backed shared memory should scale much better than System V semaphores since both of those are built on top of futex().

**Comment by mwilck** [comment]
Back then, the semaphores scaled pretty well for me. What scaled badly was fcntl() based file locking. Recent kernel patches should have improved that a lot. But anyway, I think we're moving in a different direction now, without locking. Which is good.

The point here is that you are not using any locking, and still the scaling isn't optimal. It might be helpful to add some debug output  to figure out if this is just caused by the time taken to scan through the directories, or if we are seeing lots of "races" causing repeated directory scans. I can do that if you're interested.

Do my test results match your own?

**Comment by mwilck** [comment]
This aside, the main point here is that **the race condition leading to wrong symlinks is fixed** by your patch set. So, I vote for merging this, it fixes a real bug. Performance can be worked on later if it turns out to be necessary.

**Comment by msekletar** [comment]
> Do my test results match your own?

To be honest I didn't perform any scalability testing except running the udev test suite.

**Comment by keszybz** [comment]
bionic-i386 timed out in TEST-48-START-STOP-NO-RELOAD, and bionic-amd64 in TEST-34. I'll restart them.

**Comment by keszybz** [comment]
travis timed out in `hwdb-test.sh`. That's strange: with a timeout of 270 s this shouldn't happen.

File: src/udev/udev-node.c [hunk 6]
```diff
@@ -451,8 +479,15 @@ int udev_node_add(sd_device *dev, bool apply,
...
-        FOREACH_DEVICE_DEVLINK(dev, devlink)
-                (void) link_update(dev, devlink, true);
+        FOREACH_DEVICE_DEVLINK(dev, devlink) {
+                r = link_update(dev, devlink, true);
+                if (r  == -ELOOP)
+                        /* We run update_devnode() twice, and in the first invocation, that is before update of udev
+                         * db and it could happen that two contenders are replacing each other symlinks */
+                        log_device_debug_errno(dev, r, "Failed to update device symlinks: %m");
+                else if (r < 0)
+                        log_device_error_errno(dev, r, "Failed to update device symlinks: %m");
```
Comment (keszybz) [inline src/udev/udev-node.c]: each other → each other's

```C
log_device_full_errno(dev, r == -ELOOP ? LOG_DEBUG : LOG_INFO, r,
                                   "Failed to update device symlinks: %m");
```
?
 
(And similarly below.)

File: src/udev/udev-node.c [hunk 7]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+                } else {
+                        r = node_symlink(dev, target, slink);
+                        if (r < 0) {
+                                (void) unlink(filename);
+                                break;
+                        } else if (r == 1)
+                                continue;
+                }
+
+
```
Comment (keszybz) [inline src/udev/udev-node.c:262]: Doubled newline.

File: src/udev/udev-node.c [hunk 8]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
+                r = stat(dirname, &st1);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                r = link_find_prioritized(dev, add, dirname, &target);
+                if (r < 0) {
```
Comment (keszybz) [inline src/udev/udev-node.c]: Hmm, link_find_prioritized() fails for various reasons, e.g. oom. I think this branch should only be taken on -ENOENT.

File: src/udev/udev-node.c [hunk 9]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+                        } else if (r == 1)
+                                continue;
+                }
+
+
+                r = stat(dirname, &st2);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                if (memcmp(&st1, &st2, sizeof(st1)) == 0)
```
Comment (keszybz) [inline src/udev/udev-node.c]: Hmm, this checks various stuff, incl. atime. And I'd expect that that will bump atime, so the test will not succeed. Shouldn't this use `stat_inode_unmodified()` instead?

File: src/udev/udev-event.c [hunk 10]
```diff
@@ -1041,6 +1041,10 @@ int udev_event_execute_rules(UdevEvent *event,
         if (r < 0)
                 return log_device_debug_errno(dev, r, "Failed to update database under /run/udev/data/: %m");
 
+        r = update_devnode(event);
```
Comment (poettering) [inline src/udev/udev-event.c:1047]: maybe add a comment here, why this is done again. The uninitiated onlooker might otherwise get the impression this was a copy/paste error somehow

File: src/udev/udev-node.c (hunk 9 above)
Comment (poettering) [reply src/udev/udev-node.c]: yes, please use stat_inode_unmodified()

File: src/udev/udev-node.c [hunk 11]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
```
Comment (poettering) [inline src/udev/udev-node.c:237]: no need to initialize

File: src/udev/udev-node.c [hunk 12]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
```
Comment (poettering) [inline src/udev/udev-node.c]: in newer code we declare variables in the for loop, i.e. write like this:

```c
for (unsigned i = 0; i < …
```

and remove the `int i` from the top

File: src/udev/udev-node.c [hunk 13]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+                        if (unlink(slink) == 0)
+                                (void) rmdir_parents(slink, "/");
+
+                        break;
+                } else {
+                        r = node_symlink(dev, target, slink);
+                        if (r < 0) {
+                                (void) unlink(filename);
+                                break;
+                        } else if (r == 1)
```
Comment (poettering) [inline src/udev/udev-node.c]: maybe add a brief comment what r > 0 means, i.e `/* actually created a symlink, i.e. was not already in place */`

File: src/udev/udev-node.c [hunk 14]
```diff
@@ -465,8 +500,13 @@ int udev_node_remove(sd_device *dev) {
         assert(dev);
 
         /* remove/update symlinks, remove symlinks from name index */
-        FOREACH_DEVICE_DEVLINK(dev, devlink)
-                (void) link_update(dev, devlink, false);
+        FOREACH_DEVICE_DEVLINK(dev, devlink) {
+                r = link_update(dev, devlink, false);
+                if (r  == -ELOOP)
```
Comment (poettering) [inline src/udev/udev-node.c]: spurious space

File: src/udev/udev-node.c [hunk 15]
```diff
@@ -465,8 +500,13 @@ int udev_node_remove(sd_device *dev) {
...
+                        log_device_debug_errno(dev, r, "Failed to update device symlinks: %m");
+                else if (r < 0)
+                        log_device_error_errno(dev, r, "Failed to update device symlinks: %m");
```
Comment (poettering) [inline src/udev/udev-node.c]: the handling for the specific ELOOP and the generic case are exactly the same. I think the ELOOP case should say clearly what is going on:

```c
log_device_debug_errno(dev, r, "Failed to update device symlink correctly even after %u attempts, giving up.", LINK_UPDATE_MAX_RETRIES);
```

File: src/udev/udev-node.c [hunk 16]
```diff
@@ -232,7 +228,39 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1 = {}, st2 = {};
+
```
Comment (poettering) [inline src/udev/udev-node.c:238]: i figure performance of the contended case can be improved by inserting a line like this here:

```
#define CONTENTION_SLEEP_MAX_USEC (500 * USEC_PER_MSEC)
…
usec_t max_sleep = CONTENTION_SLEEP_MAX_USEC * i / (LINK_UPDATE_MAX_RETRIES-1);
if (max_sleep > 0)
        (void) usleep(random_u64() % max_sleep);
```

i.e. sleeping by randomly increasing times, to make it less likely we collide with others. In the first iteration we'd use a sleep time of zero, i.e. not wait at all.

File: src/udev/udev-node.c [hunk 17]
```diff
@@ -99,7 +102,8 @@ static int node_symlink(sd_device *dev, const char *node, const char *slink) {
         if (rename(slink_tmp, slink) < 0) {
                 r = log_device_error_errno(dev, errno, "Failed to rename '%s' to '%s': %m", slink_tmp, slink);
                 (void) unlink(slink_tmp);
-        }
+        } else
+                r = 1;
```
Comment (poettering) [inline src/udev/udev-node.c:107]: maybe add comment `/* tell caller we actually did something, i.e. the symlink wasn't in place yet, and we created it now */`

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
Love the approach (obviously ;-)). Some minor nits.

Would be curious how performance changes with the randomized usleep() suggestion above. (and maybe even when playing around with the range of the randomized sleep. My suggestion was 500ms, but in the contended case maybe it would actually be smarter to up this to something much larger, 10s or so. Would be curious how this influences performance)

**Comment by poettering** [comment]
i didn't review the perl stuff, I am not a perlguru enough for this

File: src/udev/udev-node.c (hunk 15 above)
Comment (poettering) [reply src/udev/udev-node.c]: oh, i see now that the lines are actually different: the log level is different. didn't see that first.

But I think we should log about ELOOP at least at LOG_WARN or so, too. It's a bit fishy after all if this happens. 

Hence maybe change the ELOOP case to log_device_warning_errno() and then use the log message string i suggested?

Comment (msekletar) [reply src/udev/udev-node.c]: > But I think we should log about ELOOP at least at LOG_WARN or so, too. It's a bit fishy after all if this happens.

Hmm, but the way this works right now it happens quite often. If there are two devices competing for the same symlink  with the same priority (e.g. in the first iteration of algorithm when non-existing db entries imply priority 0 for both of them) then it is quite likely that we will hit the loop limit. Basically they could be replacing each others symlink until one runs last and wins the race. Then we write db entries, one get higher priority and potential error gets correct in the second iteration.

Comment (keszybz) [reply src/udev/udev-node.c]: Maybe make it `LOG_NOTICE` and we'll downgrade it if turns out that we see the logs in real systems?

File: src/udev/udev-node.c (hunk 9 above)
Comment (msekletar) [reply src/udev/udev-node.c]: Hmm, I think I can't really use stat_inode_umodified because that function uses `mtime` fields of stat struct. And that is actually not a field but compatibility macro defined as `#define st_mtime st_mtim.tv_sec` hence if update happens within the same second that function would not notice the change. I could have uses memcmp() just on `st_mtim` member (struct timespec) but I was sloppy and just checked if the entire structure changed for some reason.

Comment (keszybz) [reply src/udev/udev-node.c]: Looks like `stat_inode_unmodified` should be fixed to check `st_mtim` instead.

File: src/udev/udev-node.c (hunk 11 above)
Comment (msekletar) [reply src/udev/udev-node.c:237]: Sure, but...

Honestly I disagree. In C programs we tend to get bitten by uninitialised variables from time to time, not initialising also makes back-porting more error prone. Not that this matter in this specific case but in general I don't think leaving variables uninitialised is a good practise (not sure why we do this in systemd).

Moreover, in many cases we opt-in for "less clever" and sometimes slower code just to keep the source code more readable and better maintainable. Hence I don't understand why should we do such micro-optimisations when in comes to variable initialisation.

File: src/udev/udev-node.c (hunk 15 above)
Comment (msekletar) [reply src/udev/udev-node.c]: I don't understand why need to go through this exercise. You would not see it if we updated udev db before running the algorithm but @poettering didn't want this. So we run algorithm twice and in first invocations it is very likely we will hit the loop limit.

Previously, we didn't log any failures from `link_update()` now all unexpected errors will be logged normally and looping will by logged at debug level. Again, this is not a anomaly but consequence of our design decision that udev db needs to be updated "last" even though we need that data in prior algorithms.

I tried to convince @poettering that this design, in my opinion, doesn't bring anything valuable because previously we could end up with wrong symlinks, now we potentially create wrong symlinks, update db and correct them. But from the user perspective, previously and after merge of this PR, return from `udev_device_is_initialized()` is still more or less meaningless and correctness is not guaranteed.

File: src/udev/udev-node.c (hunk 11 above)
Comment (keszybz) [reply src/udev/udev-node.c:237]: Consistency is paramount in such cases. If `stat` is borked and doesn't initialize the output parameter correctly, then all calls to `stat` need to pre-initialize. We have hundreds of `stat` calls, and it generally is never initialized. The same pattern should be used here.

File: src/udev/udev-node.c (hunk 12 above)
Comment (msekletar) [reply src/udev/udev-node.c]: Actually, `i` is used to figure out whether we return `-ELOOP`

File: src/udev/udev-node.c (hunk 11 above)
Comment (msekletar) [reply src/udev/udev-node.c:237]: Actually, initialisation is needed in this case. This is because the same code handles removal of symlinks as well. Basically the link directory can get removed before we do second `stat()` and that could lead to comparison of `st1` and `st2` where `st2` contains garbage.

Comment (msekletar) [reply src/udev/udev-node.c:237]: Btw, this is not theoretical issue. If I remove it, `udev-test.pl` starts failing immediately.

File: src/udev/udev-node.c (hunk 16 above)
Comment (msekletar) [reply src/udev/udev-node.c:238]: I don't think this is necessary unless scaling proves to be a problem in practice.

**Comment by msekletar** [comment]
@poettering @keszybz thanks for the reviews so far! PTAL at the newest version.

**Comment by msekletar** [comment]
@keszybz I have no idea why hwdb-test fails. AFAICT, hwdb doesn't use any code that I've touched in this PR.

File: src/udev/udev-node.c [hunk 18]
```diff
@@ -192,7 +197,7 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
         _cleanup_free_ char *target = NULL, *filename = NULL, *dirname = NULL;
         char name_enc[PATH_MAX];
         const char *id_filename;
-        int r;
+        int i = 0, r = 0;
```
Comment (keszybz) [inline src/udev/udev-node.c]: No, please don't do unused initiailizations like that. Both `r` and `i` are unconditionally initialized before first use below.

File: src/udev/udev-node.c (hunk 11 above)
Comment (keszybz) [reply src/udev/udev-node.c:237]: ~~OK, failure to initialize `st1` is ignored in certain cases, so preinitialization is required in this case.~~

I take that back: if we fail to initialize `st1` and it remains preinitialized to 0, we know that `stat_inode_unmodified` will always return false. This means that we might just as well not do the second stat. And stat is somewhat expensive, and we shouldn't do it pointlessly. So this needs to be reworked somehow.

**Review by keszybz — CHANGES_REQUESTED** [review]

File: src/udev/udev-node.c (hunk 18 above)
Comment (msekletar) [reply src/udev/udev-node.c]: Fixed.

File: src/udev/udev-node.c (hunk 11 above)
Comment (msekletar) [reply src/udev/udev-node.c:237]: On addition first `stat()` always succeeds because we create the directory just before entering the loop. On removal, either again first `stat()` either succeeds and we attempt to do also second one (but in parallel directory might get nuked) or it fails and because of `-ENOENT` return from `link_find_prioritized()` we don't even attempt the second.

I can remove initialisation of the first one because that is safe.

**Comment by msekletar** [comment]
@keszybz updated PTAL.

I removed initialisation of `i`,`r` and `st1`.

File: src/udev/udev-node.c [hunk 19]
```diff
@@ -232,7 +229,42 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
...
+                        /* We have replaced already existing symlink, possibly there is some other device trying
+                         * to claim the same symlink. Let's do one more iteration to give us a chance to fix
+                         * the error if other device actually claims the symlink with higher priority. */
+                        continue;
+
+                r = stat(dirname, &st2);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
+
+                if (stat_inode_unmodified(&st1, &st2))
```
Comment (keszybz) [inline src/udev/udev-node.c]: Sorry, but that is still wrong (or wronger, because the issue with pointless second stat remains):
We have:
```C
struct stat st1, st2 = {};   // st1 is unitializaed
r = stat(dirname, &st1);  
if (r < 0 && errno != ENOENT) // let's say errno == ENOENT, st1 is still unitialized, `if` is not taken
                        return -errno;
... // nothing touches st1
if (stat_inode_unmodified(&st1, &st2)) // BUG: here the code uses an uninitialized variable
```

File: src/basic/stat-util.c [hunk 20]
```diff
@@ -408,7 +408,7 @@ bool stat_inode_unmodified(const struct stat *a, const struct stat *b) {
         return a && b &&
                 (a->st_mode & S_IFMT) != 0 && /* We use the check for .st_mode if the structure was ever initialized */
                 ((a->st_mode ^ b->st_mode) & S_IFMT) == 0 &&  /* same inode type */
-                a->st_mtime == b->st_mtime &&
+                (memcmp(&a->st_mtim, &b->st_mtim, sizeof(a->st_mtim)) == 0) &&
```
Comment (poettering) [inline src/basic/stat-util.c]: hmm, we can't use memcmp() for this, due to potential padding issues. i.e. there might be padding between the two fields of timespec in some arch, and we shouldn#t compare that. Hence, please change to:
```c
…
a->st_mtim.tv_sec == b->st_mtim.tv_sec &&
a->st_mtim.tv_nsec == b->st_mtim.tv_nsec &&
…
```

(and that's not just theory, there are common archs where the tv_sec fields is 32bit wide, but tv_nsec 64bit, which means 32bit padding between)

File: src/udev/udev-node.c (hunk 12 above)
Comment (poettering) [reply src/udev/udev-node.c]: good point

File: src/udev/udev-node.c (hunk 16 above)
Comment (poettering) [reply src/udev/udev-node.c:238]: you mentioned you reach the loop limit all the time? I'd call that a problem...

File: src/udev/udev-node.c [hunk 21]
```diff
@@ -232,7 +229,42 @@ static int link_update(sd_device *dev, const char *slink, bool add) {
                                 r = -errno;
                 } while (r == -ENOENT);
 
-        return r;
+        for (i = 0; i < LINK_UPDATE_MAX_RETRIES; i++) {
+                struct stat st1, st2 = {};
+
+                r = stat(dirname, &st1);
+                if (r < 0 && errno != ENOENT)
+                        return -errno;
```
Comment (poettering) [inline src/udev/udev-node.c:241]: hmm, so in contrast to what i said before, st1 needs to be initialized up top, otherwise if we really see ENOENT here, it will be uninitialized, and we'd thus compare uninitialized memory. sorry for the confusion!

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/basic/stat-util.c (hunk 20 above)
Comment (msekletar) [reply src/basic/stat-util.c]: Good point

**Comment by msekletar** [comment]
@poettering @keszybz updated.

Thank you for all the suggestions and help.

**Comment by keszybz** [comment]
@mwilck, @msekletar, good to see this done!

**Comment by mwilck** [comment]
Eventually, we had the customer test this code who originally reported this race and made me suggest #9551 in the first place. Results are mixed, unfortunately. First of all, we see 

```
[   27.101200] dl360g91 systemd-udevd[685]: sdv: Failed to update device symlinks: Too many levels of symbolic links
[   27.132792] dl360g91 systemd-udevd[625]: sdo: Failed to update device symlinks: Too many levels of symbolic links
```

This message is printed if one path looses the race for the symlink. The message "Too many levels of symbolic links", created from the -ELOOP return value from `link_update()`, is a bit misleading. Normally it indicates some sort of "symbolic link loop", but here it has a totally different meaning.

Note that this message is printed in a situation where only SCSI devices are contending for the same symlink. They all have the same priority. The issue here is, AFAICS, that every contender puts itself first in the list and will only give up if some other device claims a *higher* priority. It might be better to just use e.g. alphabetical ordering to avoid ping-pong between contenders.
